Model/ResourceModel/Product/Index/Collection.php

```diff
@@ -0,0 +1,329 @@
+            ProductIndexInterface::IS_DIRTY,
+            ['eq' => Index::DB_VALUE_BOOLEAN_TRUE]
+        );
+    }
+
+    /**
+     * Filters collection by store id
+     *
+     * @param int $storeId
+     * @return Collection
+     */
+    public function addStoreIdFilter(int $storeId)
+    {
+        return $this->addFieldToFilter(
+            ProductIndexInterface::STORE_ID,
+            ['eq' => $storeId]
+        );
+    }
+
+    /**
+     * Filters collection by product id
+     *
+     * @param int $productId
+     * @return Collection
+     */
+    public function addProductIdFilter(int $productId)
+    {
+        return $this->addFieldToFilter(
+            ProductIndexInterface::PRODUCT_ID,
+            ['eq' => $productId]
+        );
+    }
+
+    /**
+     * Filters collection by product
+     *
+     * @param ProductInterface $product
+     * @return Collection
+     */
+    public function addProductFilter(ProductInterface $product)
+    {
+        return $this->addProductIdFilter($product->getId());
+    }
+
+    /**
+     * Filters collection by id (primary key)
+     *
+     * @param int $indexId
+     * @return Collection
+     */
+    public function addIdFilter(int $indexId)
+    {
+        return $this->addFieldToFilter(
+            ProductIndexInterface::ID,
+            ['eq' => $indexId]
+        );
+    }
+
+    /**
+     * Filters collection for items that out of sync
+     *
+     * @return Collection
+     */
+    public function addOutOfSyncFilter()
+    {
+        return $this->addFieldToFilter(
+            ProductIndexInterface::IN_SYNC,
+            ['eq' => Index::DB_VALUE_BOOLEAN_FALSE]
+        );
+    }
+
+    /**
+     * Filters collection for only products that are not marked as deleted
+     *
+     * @return Collection
+     */
+    public function addNotDeletedFilter()
+    {
+        return $this->addFieldToFilter(
+            ProductIndexInterface::IS_DELETED,
+            ['eq' => Index::DB_VALUE_BOOLEAN_FALSE]
+        );
+    }
+
+    /**
+     * Filters collection for only products that are marked as deleted
+     *
+     * @return Collection
+     */
+    public function addIsDeletedFilter()
+    {
+        return $this->addFieldToFilter(
+            ProductIndexInterface::IS_DELETED,
+            ['eq' => Index::DB_VALUE_BOOLEAN_TRUE]
+        );
+    }
+
+    /**
+     * Returns the first item of the collection
+     * or null if the collection is empty
+     *
+     * @return ProductIndexInterface|null
+     * @suppress PhanTypeMismatchReturn
+     */
+    public function getOneOrNull()
+    {
+        $this->getSelect()->limit(1);
+        return $this->getSize() > 0 ? $this->getLastItem() : null; // @codingStandardsIgnoreLine
+    }
+
+    /**
+     * Marks products as deleted by given product ids and store
+     *
+     * @param array $ids
+     * @param Store $store
+     * @return int
+     */
+    public function markAsDeleted(array $ids, Store $store)
```

Comment:
These methods don't belong here. These either go some sort of service or the (not sure) repository. The `Collection` is a glorified query builder and should not know about business logic.